File: 138.Copy-List-with-Random-Pointer.py
# https://leetcode.com/problems/copy-list-with-random-pointer/description/
#
# algorithms
# Medium (25.56%)
# Total Accepted:    195.5K
# Total Submissions: 765K
# beats 100.0% of python submissions

# Definition for singly-linked list with a random pointer.
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None


class Solution(object):
    def copyRandomList(self, head):
        dic = {}
        m = n = head
        while m:  # copy node itself
            dic[m] = RandomListNode(m.label)
            m = m.next
        while n:  # link the node and copy the random pointer of the current node
            dic[n].next = dic.get(n.next)
            dic[n].random = dic.get(n.random)
            n = n.next
        return dic.get(head)

File: 138.Copy-List-with-Random-Pointer.java
// https://leetcode.com/problems/copy-list-with-random-pointer/
//
// algorithms
// Medium (26.56%)
// Total Accepted:    237,259
// Total Submissions: 893,310
// beats 83.35% of java submissions

/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;
    public Node random;

    public Node() {}

    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        
        HashMap<Node, Node> map = new HashMap<>();
        Node node = head;
        while (node != null) {
            Node tmp = new Node();
            tmp.val = node.val;
            map.put(node, tmp);
            node = node.next;
        }
        
        node = head;
        while (node != null) {
            Node nodeTmp = map.get(node);
            Node nextTmp = map.get(node.next);
            Node randomTmp = map.get(node.random);
            nodeTmp.next = nextTmp;
            nodeTmp.random = randomTmp;
            node = node.next;
        }
        
        return map.get(head);
    }
}